Ned intro revision v2 (#2293)
solo-io/gloo: docs/content/installation/gateway/kubernetes/custom_resources.md

| {{< protobuf name="gateway.solo.io.RouteTable" display="RouteTable">}} | gateway.solo.io | Child Routing object for the Gloo Gateway. |
| {{< protobuf name="gloo.solo.io.Proxy" display="Proxy">}} | gloo.solo.io | A combination of Gateway resources to be parsed by Gloo pods. |
| {{< protobuf name="gloo.solo.io.Upstream" display="Upstream">}} | gloo.solo.io | Upstreams represent destinations for routing requests. |
| {{< protobuf name="gloo.solo.io.UpstreamGroup" display="UpstreamGroup">}} | gloo.solo.io | Group multiple Upstreams and/or external endpoints to be referenced by Virtual Service(s). |
| {{< protobuf name="enterprise.gloo.solo.io.AuthConfig" display="AuthConfig">}} | enterprise.gloo.solo.io | User-facing authentication configuration referenced by Virtual Service(s). |

As a quick refresher, Gloo is deployed as pods from four different container images:

* gateway
* gloo
* discovery
* gateway-proxy/ingress-proxy

The gateway, gloo, and discovery pods act as the control plane for Gloo. The data plane is handled by the gateway-proxy/ingress-proxy pods running Envoy.

Gateway is responsible for:

> * Translating Gateway and Virtual Service Custom Resources into configuration & status on the Proxy CR

Reviewer: i forgot to include RouteTables as inputs to translation, mind adding?

Author: Updated

* Validation webhooks hit the Gateway validation server to validate a configuration before application

Gloo is responsible for:

* Translating the Proxy CR into a cached Envoy config

> * Translating Upstreams, ConfigMaps, and Endpoints into a cached Envoy config

Reviewer: let's change "a cached Envoy config" to "cached Envoy config", since the current writing implies that we have to XDS snapshot caches, when we really only have one

Reviewer: it would be even better to take the two bullet points and consolidate them to:

```suggestion
* Translating Proxy, Upstreams, UpstreamGroups, Secrets, AuthConfigs, ConfigMaps, and Endpoints into a cached Envoy config
```

Author: Updated

* Serving cached Envoy configurations via xDS
* Gloo validation server is hit by Gateway validation server to validate Proxy from Gloo's point of view

Discovery is responsible for:

> * Discovering Upstreams in the cluster

Reviewer: it also discovers functions (FDS)

Author: Updated

---
title: "Custom Resource Usage"
description: An explanation of Custom Resource usage by Gloo.
weight: 15
---

> Gloo uses native constructs in Kubernetes to store data, including ConfigMaps, Secrets, and Custom Resource Definitions (CRDs). This document is meant to summarize what CRDs are created by Gloo during installation, and how these CRDs interact with the other Gloo objects to store configuration data in the cluster.

Reviewer: technically we're talking about Custom Resources here (CRs) instead of CRDs

Author: Updated to refer to CRs instead.

---

## Overview

When Gloo is installed on Kubernetes, it creates a number of Custom Resource Definitions that Gloo can use to store data. The following table describes each Custom Resource Definition, its grouping, and its purpose.

> | Name | Grouping | Purpose |

Reviewer: it would be nice to list this in order of top down, in order of which resource references which other resources.

i.e., Gateway -> Virtual Service -> ... Proxy

Author: Updated the table based on feedback.

|------|----------|---------|

> | {{< protobuf name="enterprise.gloo.solo.io.AuthConfig" display="AuthConfig">}} | enterprise.gloo.solo.io | User-facing authentication configuration |

Reviewer: note that these are referenced by virtual services

Author: Updated.

> | {{< protobuf name="gloo.solo.io.Proxy" display="Proxy">}} | gloo.solo.io | A combination of Gateway resources to be parsed by Gloo pods. |

Reviewer: ```suggestion
| {{< protobuf name="gloo.solo.io.Proxy" display="Proxy">}} | gloo.solo.io | The derived proxy configuration to be sent to Envoy. Derived from a combination of Gateway, and Ingress resources. |
```

> | {{< protobuf name="gloo.solo.io.Settings" display="Settings">}} | gloo.solo.io | Global settings for all Gloo container types. |

Reviewer: nit, prefer brevity
```suggestion
| {{< protobuf name="gloo.solo.io.Settings" display="Settings">}} | gloo.solo.io | Global settings for all Gloo containers. |
```

> | {{< protobuf name="gloo.solo.io.UpstreamGroup" display="UpstreamGroup">}} | gloo.solo.io | Defining multiple Upstreams or external endpoints for a Virtual Service. |

Reviewer: ```suggestion
| {{< protobuf name="gloo.solo.io.UpstreamGroup" display="UpstreamGroup">}} | gloo.solo.io | Group multiple Upstreams and/or external endpoints to be referenced by Virtual Service(s). |
```

Important to highlight the main benefit of this is referencing the same upstream group from several virtual services

| {{< protobuf name="gloo.solo.io.Upstream" display="Upstream">}} | gloo.solo.io | Upstreams represent destinations for routing requests. |

> | {{< protobuf name="gateway.solo.io.Gateway" display="Gateway">}} | gateway.solo.io | Describes a single Listener and the routing Upstreams reachable via the Gateway Proxy. |

Reviewer: we should note that gateway listeners reference virtual services

Author: Updated

| {{< protobuf name="gateway.solo.io.RouteTable" display="RouteTable">}} | gateway.solo.io | Child Routing object for the Gloo Gateway. |

> | {{< protobuf name="gateway.solo.io.VirtualService" display="VirtualService">}} | gateway.solo.io | Describes the set of routes to match for a set of domains. |

Reviewer: we should note that virtual services reference route tables & upstreams/upstream groups

Author: Updated

> As a quick refresher, Gloo is deployed from several different container images. The Gateway image is responsible for acting as the API gateway and routing traffic to the appropriate Envoy Proxy. The Gloo image is the control plane for Gloo and runs the xDS server used to interact with the Envoy Proxy instances. The Discovery image performs service discovery of new services added to Kubernetes or another plugin.

Reviewer: > The Gateway image is responsible for acting as the API gateway and routing traffic to the appropriate Envoy Proxy. The Gloo image is the control plane for Gloo

this isn't exactly correct.

an open-source gloo installation comes with four pods:
- gateway
- gloo
- discovery
- gateway-proxy/ingress-proxy(ies)

The control plane consists of the first three, and the data-plane is envoy (gateway-proxy/ingress proxy(ies))

Gateway is responsible for:
- translating gateway and virtual service custom resources into configuration & status on the Proxy CR
- validation webhook hits gateway validation server to validate config before application

Gloo is responsible for:
- translating Proxy CR into cached Envoy config
- translating upstreams, configmaps, endpoints, into cached Envoy config
- serving cached Envoy config via XDS
- gloo validation server is hit by gateway validation server to validate proxy from Gloo's point of view

Discovery is responsible for:
- discovering upstreams in the cluster

Gateway-Proxy/Ingress-Proxy:
- Is Envoy

Author: Updated to reflect this information.
